src/artifacts.js

```diff
@@ -423,6 +643,99 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
+  // marked as present short circuit
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
+  // Authenticate request by providing parameters
+  if (!req.satisfies({
+    taskId,
+    runId,
+    workerGroup,
+    workerId,
+    name,
+  })) {
+    return;
+  }
+
+  if (artifact.storageType === 'blob') {
+    let etag;
+    if (artifact.details.partsHash) {
+      etag = await this.s3Controller.completeMultipartUpload({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        etags: input.etags,
+        uploadId: artifact.details.uploadId,
+        tags: {taskId, runId, name},
+      });
+    } else {
+      etag = input.etags[0];
```

Comment:
This is where we're supposed to do a `HEAD` request, otherwise the `artifactCreated` message has no value...

In fact we don't promise that artifacts are present if we don't this here....